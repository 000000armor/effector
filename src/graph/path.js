//@noflow

import type {Derivable} from '../derive'
import {isDerivable} from '../derive/types'
import * as Kind from '../kind'
import type {Store} from '../store/index.h'
export type RootValue = -2
export type OpaqueValue = -1
export type NoValue = 0
export type PlainValue = 1
export type DerivedValue = 2
export type StoreValue = 3
export type StoreObjectValue = 4
export type StoreArrayValue = 5

// type RootNode = 1
// type MiddleNode = 2
// type NodeType = RootNode | MiddleNode
export type Segment =
 | [PlainValue, void]
 | [RootValue, () => any]
 | [DerivedValue, void]
 | [NoValue, void]
 | [StoreValue, void]
 | [StoreObjectValue, string]
 | [StoreArrayValue, number]
type ValueElement = Segment | [OpaqueValue, any]

type Path = ValueElement[]
type StateTree = [ValueElement, Path]

function goDeep(
 nextElement: ValueElement,
 [element, path]: StateTree,
): StateTree {
 return [nextElement, [...path, element]]
}
function flatten([element, path]: StateTree): Path {
 return [...path, element] //.filter(e => e.type !== -1)
}

function storeType(thing: any, stateTree: StateTree): StateTree[] {
 const currentTree = stateTree /* goDeep(
  {
   type: (-1: OpaqueValue),
   value: thing,
  },
  stateTree,
 )*/
 if (thing === undefined || thing === null)
  return [goDeep([(0: NoValue) /*::, undefined*/], currentTree)]
 if (isDerivable(thing)) {
  return storeTypeDerive(thing, currentTree)
 }
 if (Kind.isStore(thing)) {
  return storeTypeStore(thing, currentTree)
 }
 if (Array.isArray(thing)) {
  return storeTypeArray(thing, currentTree)
 }
 if (thing.constructor === Object) {
  return storeTypeObject(thing, currentTree)
 }
 return [goDeep([(1: PlainValue) /*::, undefined*/], stateTree)]
}
function storeTypeDerive(thing: Derivable<any>, stateTree: StateTree) {
 return storeType(
  thing.get(),
  goDeep([(2: DerivedValue) /*::, undefined*/], stateTree),
 )
}
function storeTypeStore(thing: Store<any>, stateTree: StateTree) {
 return storeType(
  thing.stateAtom.get(),
  goDeep([(3: StoreValue) /*::, undefined*/], stateTree),
 )
}
function storeTypeObject(thing: Object, stateTree: StateTree) {
 const childrens: StateTree[] = []

 for (const field of Object.keys(thing)) {
  const e = thing[field]
  const nested = storeType(e, goDeep([(4: StoreObjectValue), field], stateTree))
  childrens.push(...nested)
 }

 return childrens
}

function storeTypeArray(thing: Array<any>, stateTree: StateTree) {
 const childrens: StateTree[] = []

 for (let i = 0; i < thing.length; i++) {
  const e = thing[i]
  const nested = storeType(e, goDeep([(5: StoreArrayValue), i], stateTree))
  childrens.push(...nested)
 }

 return childrens
}

// function storeFabric<T>({defaultValue, ...opts}: {defaultValue: T}) {
//  // const isPlainStore =
//  const value: Derivable<T> = derive(() => deepUnpack(arg))
//  const proxyValue = lens({
//   get() {
//    return value.get()
//   },
//  })
//  const store = {
//   kind: Kind.STORE,
//  }
//  return store
// }
function readSegment(current: any, segment: Segment) {
 switch (segment[0]) {
  case (-2: RootValue):
   return segment[1]()
  case (0: NoValue):
   return null
  case (1: PlainValue):
   return current
  case (2: DerivedValue):
   return current.get()
  case (3: StoreValue):
   return current.stateAtom.get()
  case (4: StoreObjectValue):
   return current[segment[1]]
  case (5: StoreArrayValue):
   return current[segment[1]]
  default:
   throw new TypeError(`Wrong segment ${segment.toString()}`)
 }
}

export function readPath(path: Path): any {
 let current = null
 for (const e of path) {
  if (e[0] !== (-1: OpaqueValue)) current = readSegment(current, e)
 }
 return current
}
export function getStorePath(thing: () => any): Path[] {
 return storeType(thing(), [[(-2: RootValue), thing], []]).map(flatten)
}
