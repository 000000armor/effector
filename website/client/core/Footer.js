/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

const React = require('react')

class Footer extends React.Component {
  docUrl(doc, language) {
    const baseUrl = this.props.config.baseUrl
    const docsUrl = this.props.config.docsUrl
    const docsPart = `${docsUrl ? `${docsUrl}/` : ''}`
    return `${baseUrl}${docsPart}${doc}`
  }

  pageUrl(doc, language) {
    return `${this.props.config.baseUrl}${doc}`
  }

  render() {
    return (
      <footer className="nav-footer" id="footer">
        <section className="sitemap">
          <a href={this.props.config.baseUrl} className="nav-home">
            {this.props.config.footerIcon && (
              <img
                src={this.props.config.baseUrl + this.props.config.footerIcon}
                alt={this.props.config.title}
              />
            )}
          </a>
          <div>
            <h5>Docs</h5>
            <a href={this.docUrl('introduction/installation')}>
              Getting Started
            </a>
            <a href={this.docUrl('api/effector/effector')}>API Reference</a>
            <a href="https://changelog.effector.dev">Changelog</a>
          </div>
          <div>
            <h5>Community</h5>
            <a
              href="https://twitter.com/effectorjs"
              target="_blank"
              rel="noreferrer noopener">
              Twitter
            </a>
            <a href="https://t.me/effector_ru">Telegram ðŸ‡·ðŸ‡º</a>
            <a href="https://t.me/effector_en">Telegram</a>
          </div>
          <div>
            <h5>More</h5>
            <a href="https://github.com/zerobias/effector">GitHub</a>
            <a
              className="github-button"
              href={this.props.config.repoUrl}
              data-icon="octicon-star"
              data-count-href="/zerobias/effector/stargazers"
              data-show-count="true"
              data-count-aria-label="# stargazers on GitHub"
              aria-label="Star this project on GitHub">
              Star
            </a>
          </div>
        </section>
        <section className="copyright">{this.props.config.copyright}</section>
      </footer>
    )
  }
}

module.exports = Footer
