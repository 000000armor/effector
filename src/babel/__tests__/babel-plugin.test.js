//@flow strict

import fs from 'fs'
import path from 'path'
import {transformFileSync} from '@babel/core'

describe('babel-plugin', () => {
  const fixturesDir = path.join(__dirname, 'fixtures')
  const testCases = fs
    .readdirSync(fixturesDir)
    .filter(file => file.endsWith('.js'))
    .sort()
  for (const caseFile of testCases) {
    const caseName = caseFile
      .split('-')
      .join(' ')
      .slice(0, -3)

    const optionsName = `${caseFile.slice(0, -3)}.options.json`
    const optionsPath = path.join(fixturesDir, optionsName)
    const hasOptions = fs.existsSync(optionsPath)

    if (hasOptions) {
      const options = JSON.parse(
        fs.readFileSync(optionsPath, {encoding: 'utf8'}).toString(),
      )

      it(`should ${caseName} with options`, () => {
        const fixturePath = path.join(fixturesDir, caseFile)
        const fixture = transformFileSync(fixturePath, {
          // configFile: path.resolve(__dirname, '../../../babel.config.js'),
          configFile: false,
          babelrc: false,
          envName: 'test',
          plugins: [[path.resolve(__dirname, '../babel-plugin.js'), options]],
        })?.code

        expect(fixture).toMatchSnapshot()
      })
    } else {
      it(`should ${caseName}`, () => {
        const fixturePath = path.join(fixturesDir, caseFile)
        const fixture = transformFileSync(fixturePath, {
          configFile: path.join(__dirname, '.babelrc'),
        })?.code

        expect(fixture).toMatchSnapshot()
      })
    }
  }
})
