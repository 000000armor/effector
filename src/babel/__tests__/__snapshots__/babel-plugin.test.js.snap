// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-plugin should attach 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { attach, createEffect } from 'effector';
const effect = createEffect({
  name: \\"effect\\",
  sid: \\"cqzl92\\"
});
const foo1 = attach({
  ɔ: {
    effect,
    mapParams: _ => _
  },
  config: {
    name: \\"foo1\\",
    sid: \\"-ko4avk\\"
  }
});
attach({
  ɔ: {
    effect,
    mapParams: _ => _
  },
  config: {
    sid: \\"tzqz4n\\"
  }
});
const config = {
  effect,
  mapParams: _ => _
};
const foo2 = attach({
  ɔ: config,
  config: {
    name: \\"foo2\\",
    sid: \\"-4t7i6y\\"
  }
});

const f = () => attach({
  ɔ: {
    effect,
    mapParams: _ => _
  },
  config: {
    name: \\"f\\",
    sid: \\"hf8d25\\"
  }
});"
`;

exports[`babel-plugin should combine 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createStore, combine } from 'effector';
const foo = createStore('foo', {
  name: \\"foo\\",
  sid: \\"vir1ix\\"
});
const bar = createStore('h', {
  name: \\"bar\\",
  sid: \\"-92kryt\\"
});
const baz1 = combine({
  ɔ: [{
    foo,
    bar
  }],
  config: {
    name: \\"baz1\\",
    sid: \\"ihqrp\\"
  }
});
const baz2 = combine({
  ɔ: [[foo, bar]],
  config: {
    name: \\"baz2\\",
    sid: \\"-2zo8sd\\"
  }
});
const baz3 = combine({
  ɔ: [foo, bar, (a, b) => [a, b]],
  config: {
    name: \\"baz3\\",
    sid: \\"-6hu8cf\\"
  }
});
const spread = [foo, bar];
const baz4 = combine({
  ɔ: [...spread],
  config: {
    name: \\"baz4\\",
    sid: \\"-cwioto\\"
  }
});
const shape = {
  foo,
  bar
};
const baz5 = combine({
  ɔ: [shape],
  config: {
    name: \\"baz5\\",
    sid: \\"79adte\\"
  }
});
combine({
  ɔ: [{
    foo,
    bar
  }],
  config: {
    sid: \\"6d0wwi\\"
  }
});
combine({
  ɔ: [...spread],
  config: {
    sid: \\"6dkphv\\"
  }
});

const f = a => combine({
  ɔ: [{
    foo,
    bar
  }],
  config: {
    name: \\"f\\",
    sid: \\"-arx9a6\\"
  }
});"
`;

exports[`babel-plugin should create component 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createStore } from 'effector';
import { createComponent } from 'effector-react';
const foo = createStore('foo', {
  name: \\"foo\\",
  sid: \\"-jdfd95\\"
});
const Foo = createComponent(foo, (props, data) => null, {
  loc: {
    file: _effectorFileName,
    line: 6,
    column: 12
  },
  name: \\"Foo\\",
  sid: \\"Foo /src/babel/__tests__/fixtures/create-component.js [6, 12]\\"
});"
`;

exports[`babel-plugin should create domain 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createDomain } from 'effector';
const domain = createDomain({
  name: \\"domain\\",
  sid: \\"-r2jo9q\\"
});
const foo = createDomain('bar', {
  name: \\"foo\\",
  sid: \\"-kylwgu\\"
});
const {
  sid
} = createDomain({
  sid: \\"-ikueb0\\"
});
const {
  name
} = createDomain('foo', {
  sid: \\"-i3srvy\\"
});
const {
  shortName
} = createDomain({
  name: 'foo'
}, {
  sid: \\"-hmr4vd\\"
});
createDomain({
  sid: \\"-klzve2\\"
});
createDomain({
  sid: \\"-kkwa7c\\"
});"
`;

exports[`babel-plugin should create domain domain 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createDomain } from 'effector';
const domain = createDomain({
  name: \\"domain\\",
  sid: \\"-jsqfil\\"
});
const foo = domain.domain({
  name: \\"foo\\",
  sid: \\"-veeiq1\\"
});
const bar = domain.domain('baz', {
  name: \\"bar\\",
  sid: \\"ksdvxf\\"
});
const fooAlias = domain.createDomain({
  name: \\"fooAlias\\",
  sid: \\"75cvir\\"
});
const barAlias = domain.createDomain('bazAlias', {
  name: \\"barAlias\\",
  sid: \\"-abn323\\"
});
const {
  sid
} = domain.createDomain({
  sid: \\"sm28rc\\"
});
const {
  name
} = domain.createDomain('foo', {
  sid: \\"t33v6e\\"
});
const {
  shortName
} = domain.createDomain({
  name: 'foo'
}, {
  sid: \\"tk5i6z\\"
});
domain.createDomain({
  sid: \\"asytpf\\"
});
domain.createDomain({
  sid: \\"au2ew5\\"
});"
`;

exports[`babel-plugin should create domain effect 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createDomain } from 'effector';
const domain = createDomain({
  name: \\"domain\\",
  sid: \\"-gfxji2\\"
});
const a = domain.effect({
  name: \\"a\\",
  sid: \\"-yaax6i\\"
});

const f = () => domain.effect({
  name: \\"f\\",
  sid: \\"jyl44k\\"
});

const b = domain.effect('hello', {
  name: \\"b\\",
  sid: \\"y8n8xx\\"
});

const g = () => domain.effect('baz', {
  name: \\"g\\",
  sid: \\"hgf89v\\"
});

const aAlias = domain.createEffect({
  name: \\"aAlias\\",
  sid: \\"-ej7o7u\\"
});

const fAlias = () => domain.createEffect({
  name: \\"fAlias\\",
  sid: \\"-yfer5d\\"
});

const bAlias = domain.createEffect('helloAlias', {
  name: \\"bAlias\\",
  sid: \\"-hnrrtn\\"
});

const gAlias = () => domain.createEffect('bazAlias', {
  name: \\"gAlias\\",
  sid: \\"xh577y\\"
});

const {
  sid
} = domain.createEffect({
  sid: \\"-7j3t5r\\"
});
const {
  name
} = domain.createEffect('foo', {
  sid: \\"-7226qp\\"
});
const {
  shortName
} = domain.createEffect({
  name: 'foo'
}, {
  sid: \\"-6l0jq4\\"
});
domain.createEffect({
  sid: \\"e7yw3e\\"
});
domain.createEffect({
  sid: \\"ekm5qp\\"
});"
`;

exports[`babel-plugin should create domain event 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createDomain } from 'effector';
const domain = createDomain({
  name: \\"domain\\",
  sid: \\"3a2uet\\"
});
const a = domain.event({
  name: \\"a\\",
  sid: \\"-h01fmt\\"
});

const f = () => domain.event({
  name: \\"f\\",
  sid: \\"-nyhxip\\"
});

const b = domain.event('asm', {
  name: \\"b\\",
  sid: \\"p2tf20\\"
});

const g = () => domain.event('tasm', {
  name: \\"g\\",
  sid: \\"i4cx64\\"
});

const aAlias = domain.createEvent({
  name: \\"aAlias\\",
  sid: \\"ocb24t\\"
});

const fAlias = () => domain.createEvent({
  name: \\"fAlias\\",
  sid: \\"sqoaj0\\"
});

const bAlias = domain.createEvent('asmAlias', {
  name: \\"bAlias\\",
  sid: \\"-2c6lli\\"
});

const gAlias = () => domain.createEvent('tasmAlias', {
  name: \\"gAlias\\",
  sid: \\"226msp\\"
});

const {
  sid
} = domain.createEvent({
  sid: \\"68j0kq\\"
});
const {
  name
} = domain.createEvent('foo', {
  sid: \\"6pkmzs\\"
});
const {
  shortName
} = domain.createEvent({
  name: 'foo'
}, {
  sid: \\"76ma0d\\"
});
domain.createEvent({
  sid: \\"-v9zudr\\"
});
domain.createEvent({
  sid: \\"-uxckqg\\"
});
domain.createEvent('foo', {
  sid: \\"-uw8zjq\\"
});"
`;

exports[`babel-plugin should create domain store 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createDomain } from 'effector';
const domain = createDomain({
  name: \\"domain\\",
  sid: \\"-fpi1bm\\"
});
const a = domain.store('bar', {
  name: \\"a\\",
  sid: \\"z1hqlw\\"
});
const b = domain.store('h', {
  ɔ: {
    option: 'test',
    ['na' + 'me']: 'LOL'
  },
  name: \\"b\\",
  sid: \\"5m6wxe\\"
});
const c = domain.store('h', {
  ɔ: {
    name: 'test'
  },
  name: \\"c\\",
  sid: \\"-nt3wr4\\"
}); //$off

const d = domain.store('h', {
  ɔ: null,
  name: \\"d\\",
  sid: \\"a6fhta\\"
}); //$off

const e = domain.store('h', {
  ɔ: 4234,
  name: \\"e\\",
  sid: \\"7o9lyl\\"
});
const f = domain.store('h', {
  ɔ: {},
  name: \\"f\\",
  sid: \\"4p23pp\\"
});
const aAlias = domain.createStore('bar', {
  name: \\"aAlias\\",
  sid: \\"6bga5r\\"
});
const bAlias = domain.createStore('h', {
  ɔ: {
    option: 'test',
    ['na' + 'me']: 'LOL'
  },
  name: \\"bAlias\\",
  sid: \\"-ku2zyr\\"
});
const cAlias = domain.createStore('h', {
  ɔ: {
    name: 'testAlias'
  },
  name: \\"cAlias\\",
  sid: \\"n1hrvv\\"
}); //$off

const dAlias = domain.createStore('h', {
  ɔ: null,
  name: \\"dAlias\\",
  sid: \\"-3mzvug\\"
}); //$off

const eAlias = domain.createStore('h', {
  ɔ: 4234,
  name: \\"eAlias\\",
  sid: \\"-kdjhao\\"
});
const fAlias = domain.createStore('h', {
  ɔ: {},
  name: \\"fAlias\\",
  sid: \\"ni1ajy\\"
});
const {
  sid
} = domain.createStore(null, {
  sid: \\"-f8txo\\"
});
const {
  shortName
} = domain.createStore(null, {
  ɔ: {
    name: 'foo'
  },
  sid: \\"1st3s\\"
});
domain.createStore(null, {
  sid: \\"l6xkh2\\"
});
domain.createStore(null, {
  sid: \\"l815ns\\"
});"
`;

exports[`babel-plugin should create effect 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createEffect } from 'effector';
const foo = createEffect({
  name: \\"foo\\",
  sid: \\"-hlt0gb\\"
});
const bar = createEffect('hello', {
  name: \\"bar\\",
  sid: \\"-ydp1p5\\"
});
const baz = createEffect({
  handler() {
    return 0;
  }

}, {
  name: \\"baz\\",
  sid: \\"-v7s7ya\\"
});
const quux = createEffect('nice', {
  ɔ: {
    handler() {
      return 0;
    }

  },
  name: \\"quux\\",
  sid: \\"-evyqqj\\"
});

const f = () => createEffect({
  name: \\"f\\",
  sid: \\"y82cl8\\"
});

const {
  sid
} = createEffect({
  sid: \\"el3ns6\\"
});
const {
  name
} = createEffect('foo', {
  sid: \\"p03chb\\"
});
const {
  shortName
} = createEffect({
  name: 'foo'
}, {
  sid: \\"ph4zhw\\"
});
createEffect({
  sid: \\"-gr1rsm\\"
});
createEffect({
  sid: \\"-gpy6lw\\"
});"
`;

exports[`babel-plugin should create event 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createEvent } from 'effector';
const foo = createEvent({
  name: \\"foo\\",
  sid: \\"qm3jl6\\"
});
const bar = createEvent('hello', {
  name: \\"bar\\",
  sid: \\"f2obva\\"
});
const baz = createEvent({
  name: 'nice'
}, {
  name: \\"baz\\",
  sid: \\"-yrvnsz\\"
});

const f = () => createEvent({
  name: \\"f\\",
  sid: \\"vi2m03\\"
});

const {
  sid
} = createEvent({
  sid: \\"-bimhsq\\"
});
const {
  name
} = createEvent('foo', {
  sid: \\"-b1kvdo\\"
});
const {
  shortName
} = createEvent({
  name: 'foo'
}, {
  sid: \\"-akj8d3\\"
});
createEvent({
  sid: \\"-4vb1bf\\"
});
createEvent({
  sid: \\"-4u7g4p\\"
});"
`;

exports[`babel-plugin should create store 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createStore } from 'effector';
const foo = createStore('foo', {
  name: \\"foo\\",
  sid: \\"75h1gk\\"
});
const a = createStore('h', {
  name: \\"a\\",
  sid: \\"osbuqo\\"
});
const b = createStore('h', {
  ɔ: {},
  name: \\"b\\",
  sid: \\"-yawrls\\"
}); //$off

const c = createStore('h', {
  ɔ: 23020,
  name: \\"c\\",
  sid: \\"-lvzpkx\\"
});
const config = {
  option: 0
};
const dod = createStore(null, {
  ɔ: config,
  name: \\"dod\\",
  sid: \\"qtb6qu\\"
});

const f = a => createStore(a, {
  name: \\"f\\",
  sid: \\"42loz8\\"
});

const {
  sid
} = createStore(null, {
  sid: \\"-u0iw5z\\"
});
const {
  shortName
} = createStore(null, {
  ɔ: {
    name: 'foo'
  },
  sid: \\"-tjh94j\\"
});
createStore(null, {
  sid: \\"-nt8j9r\\"
});
createStore(null, {
  sid: \\"-ns4y31\\"
});"
`;

exports[`babel-plugin should createApi 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createStore, createEvent, createApi } from 'effector';
const foo = createStore([], {
  name: \\"foo\\",
  sid: \\"-hmtdwm\\"
});
const bar = createEvent({
  name: \\"bar\\",
  sid: \\"bo1nrg\\"
});
const baz1 = createApi({
  ɔ: [foo, {
    reset: () => []
  }],
  config: {
    sid: \\"i3pgik\\"
  }
});
createApi({
  ɔ: [foo, {
    reset: () => []
  }],
  config: {
    sid: \\"-3zehg3\\"
  }
});
const spread = [foo, {
  reset: () => []
}];
const baz4 = createApi({
  ɔ: [...spread],
  config: {
    sid: \\"-ljq5ad\\"
  }
});

const f = () => createApi({
  ɔ: [foo, {
    reset: () => []
  }],
  config: {
    sid: \\"-klmwfe\\"
  }
});"
`;

exports[`babel-plugin should forward 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createStore, createEvent, forward } from 'effector';
const foo = createStore('foo', {
  name: \\"foo\\",
  sid: \\"c0eo5u\\"
});
const bar = createEvent({
  name: \\"bar\\",
  sid: \\"-skx5bw\\"
});
const baz1 = forward({
  ɔ: {
    from: foo,
    to: bar
  },
  config: {
    name: \\"baz1\\",
    sid: \\"-izumle\\"
  }
});
const baz2 = forward({
  ɔ: {
    from: foo,
    to: bar
  },
  config: {
    name: \\"baz2\\",
    sid: \\"-mi0m5g\\"
  }
});
forward({
  ɔ: {
    from: foo,
    to: bar
  },
  config: {
    sid: \\"-do0gmv\\"
  }
});
forward({
  ɔ: {
    from: foo,
    to: bar
  },
  config: {
    sid: \\"-dngo1i\\"
  }
});
const config = {
  from: foo,
  to: bar
};
const baz5 = forward({
  ɔ: config,
  config: {
    name: \\"baz5\\",
    sid: \\"-tagonb\\"
  }
});

const f = () => forward({
  ɔ: {
    from: foo,
    to: bar
  },
  config: {
    name: \\"f\\",
    sid: \\"7acgzo\\"
  }
});"
`;

exports[`babel-plugin should guard 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createStore, createEvent, guard } from 'effector';
const foo = createStore(true, {
  name: \\"foo\\",
  sid: \\"12fz6b\\"
});
const bar = createEvent({
  name: \\"bar\\",
  sid: \\"-3ob7uj\\"
});
const baz1 = guard({
  ɔ: [bar, {
    filter: foo
  }],
  config: {
    name: \\"baz1\\",
    sid: \\"b10z6n\\"
  }
});
const baz2 = guard({
  ɔ: [{
    source: bar,
    filter: foo
  }],
  config: {
    name: \\"baz2\\",
    sid: \\"rjc331\\"
  }
});
guard({
  ɔ: [bar, {
    filter: foo
  }],
  config: {
    sid: \\"-ajecgo\\"
  }
});
guard({
  ɔ: [{
    source: bar,
    filter: foo
  }],
  config: {
    sid: \\"-aiujvb\\"
  }
});
const spread = [bar, {
  filter: foo
}];
const baz4 = guard({
  ɔ: [...spread],
  config: {
    name: \\"baz4\\",
    sid: \\"-dxz69v\\"
  }
});
const config = {
  source: bar,
  filter: foo
};
const baz5 = guard({
  ɔ: [config],
  config: {
    name: \\"baz5\\",
    sid: \\"-ara4o9\\"
  }
});

const f = () => guard({
  ɔ: [{
    source: bar,
    filter: foo
  }],
  config: {
    name: \\"f\\",
    sid: \\"-7ro5ad\\"
  }
});"
`;

exports[`babel-plugin should imports 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createStore as createStoreEffector } from 'effector';
import { createStore } from 'redux';
const x = createStoreEffector(0, {
  name: \\"x\\",
  sid: \\"-x1t9c8\\"
});
const e = createEvent({
  name: \\"e\\",
  sid: \\"-s0v3ue\\"
});
const reduxStore = createStore(() => 0, {
  name: \\"reduxStore\\",
  sid: \\"-mmcts9\\"
});"
`;

exports[`babel-plugin should restore 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createEvent, restore } from 'effector';
const foo = createEvent({
  name: \\"foo\\",
  sid: \\"uytd56\\"
});
const a = restore(foo, null, {
  name: \\"a\\",
  sid: \\"-n6x72o\\"
});
const b = restore(foo, null, {
  ɔ: {},
  name: \\"b\\",
  sid: \\"-qp36mq\\"
}); //$off

const c = restore(foo, null, {
  ɔ: 23020,
  name: \\"c\\",
  sid: \\"-tq7jsl\\"
});
const config = {
  option: 0
};
const dod = restore(foo, null, {
  ɔ: config,
  name: \\"dod\\",
  sid: \\"-5s8kqi\\"
});

const f = a => restore(foo, null, {
  name: \\"f\\",
  sid: \\"tcvrf2\\"
});

const {
  sid
} = restore(foo, null, {
  sid: \\"duje3b\\"
});
const {
  shortName
} = restore(foo, null, {
  ɔ: {
    name: 'bar'
  },
  sid: \\"ebl14r\\"
});
restore(foo, null, {
  sid: \\"5t38ir\\"
});
restore(foo, null, {
  sid: \\"5u6tph\\"
});"
`;

exports[`babel-plugin should sample 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createStore, createEvent, sample } from 'effector';
const foo = createStore('foo', {
  name: \\"foo\\",
  sid: \\"-4e9wqe\\"
});
const bar = createEvent({
  name: \\"bar\\",
  sid: \\"-nktibu\\"
});
const baz1 = sample({
  ɔ: [foo, bar],
  config: {
    name: \\"baz1\\",
    sid: \\"-mcp4ju\\"
  }
});
const baz2 = sample({
  ɔ: [{
    source: foo,
    clock: bar
  }],
  config: {
    name: \\"baz2\\",
    sid: \\"-m09tbe\\"
  }
});
sample({
  ɔ: [foo, bar],
  config: {
    sid: \\"u6o7kj\\"
  }
});
sample({
  ɔ: [{
    source: foo,
    clock: bar
  }],
  config: {
    sid: \\"u7805w\\"
  }
});
const spread = [foo, bar];
const baz4 = sample({
  ɔ: [...spread],
  config: {
    name: \\"baz4\\",
    sid: \\"55gj6o\\"
  }
});
const config = {
  source: foo,
  clock: bar
};
const baz5 = sample({
  ɔ: [config],
  config: {
    name: \\"baz5\\",
    sid: \\"4hk9zs\\"
  }
});

const f = () => sample({
  ɔ: [{
    source: foo,
    clock: bar
  }],
  config: {
    name: \\"f\\",
    sid: \\"hzwgc2\\"
  }
});"
`;

exports[`babel-plugin should split 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createStore, createEvent, split } from 'effector';
const foo = createStore([], {
  name: \\"foo\\",
  sid: \\"-pyo7gi\\"
});
const bar = createEvent({
  name: \\"bar\\",
  sid: \\"-upfehc\\"
});
const baz1 = split({
  ɔ: [foo, {
    nonEmpty: list => list.length === 0
  }],
  config: {
    sid: \\"-yuao1s\\"
  }
});
split({
  ɔ: [foo, {
    nonEmpty: list => list.length === 0
  }],
  config: {
    sid: \\"-13wt2v\\"
  }
});
const baz2 = split({
  ɔ: [{
    source: foo,
    match: {
      nonEmpty: list => list.length === 0
    },
    cases: {
      nonEmpty: bar
    }
  }],
  config: {
    sid: \\"-xw7f9e\\"
  }
});
split({
  ɔ: [{
    source: foo,
    match: {
      nonEmpty: list => list.length === 0
    },
    cases: {
      nonEmpty: bar
    }
  }],
  config: {
    sid: \\"xhp42d\\"
  }
});
const spread = [foo, {
  nonEmpty: list => list.length === 0
}];
const baz4 = split({
  ɔ: [...spread],
  config: {
    sid: \\"-dyswqx\\"
  }
});
const config = {
  source: foo,
  match: {
    nonEmpty: list => list.length === 0
  },
  cases: {
    nonEmpty: bar
  }
};
const baz5 = split({
  ɔ: [config],
  config: {
    sid: \\"-8hvba\\"
  }
});

const f = () => split({
  ɔ: [foo, {
    nonEmpty: list => list.length === 0
  }],
  config: {
    sid: \\"pldjp\\"
  }
});"
`;

exports[`babel-plugin should use store 1`] = `
"var _effectorFileName = \\"\\";
//@flow
import { createStore } from 'effector';
import { useStore } from 'effector-react';
const foo = createStore('foo', {
  name: \\"foo\\",
  sid: \\"ciaon0\\"
});
const bar = createStore('bar', {
  name: \\"bar\\",
  sid: \\"-t7ybo2\\"
});

const Foo = () => {
  const data = useStore(foo, {
    loc: {
      file: _effectorFileName,
      line: 8,
      column: 15
    },
    name: \\"Foo\\",
    sid: \\"Foo /src/babel/__tests__/fixtures/use-store.js [8, 15]\\"
  });
  const data0 = useStore(bar, {
    loc: {
      file: _effectorFileName,
      line: 9,
      column: 16
    },
    name: \\"Foo\\",
    sid: \\"Foo /src/babel/__tests__/fixtures/use-store.js [9, 16]\\"
  });
};"
`;
